### PR TITLE
LIMS-4269: edit suborder test unit

### PR DESCRIPTION
**Pull request by ghada-37**
**When I delete test unit to update it message will appear ( This Test Unit will be removed from the corresponding analysis )
-Make sure the corresponding analysis records created according to this update in  test unit.**

```
➜  1lims-automation git:(LIMS-4269-testunits) ✗ nosetests -vs --nologcapture --tc-file=config.ini ui_testing/testcases/basic_tests/test06_orders.py:OrdersTestCases.test022_update_suborder_testunits
2020-05-28 04:09:26.021 | INFO     | api_testing.apis.base_api:info:76 - Get authorized api session.
2020-05-28 04:09:26.696 | INFO     | api_testing.apis.base_api:info:76 - session ID : eyJhbGciOi .....
-When I delete test unit to update it message will appear ...   
2020-05-28 04:09:26.866 | INFO     | ui_testing.testcases.base_test:info:129 - Test case : test022_update_suborder_testunits
2020-05-28 04:09:43.626 | INFO     | ui_testing.pages.base_pages:wait_until_page_is_loaded:531 - wait until page is loaded
2020-05-28 04:09:44.908 | INFO     | api_testing.apis.base_api:info:76 - GET : https://automation.1lims.com/api/orders
2020-05-28 04:09:46.098 | INFO     | api_testing.apis.base_api:info:76 - Status code: 1
2020-05-28 04:09:46.099 | INFO     | api_testing.apis.base_api:info:76 - GET : https://automation.1lims.com/api/orders/get/subOrder/byMainOrderId?orderId=1580&deleted=0
2020-05-28 04:09:46.417 | INFO     | api_testing.apis.base_api:info:76 - Status code: 1
2020-05-28 04:09:46.418 | INFO     | api_testing.apis.base_api:info:76 - GET : https://automation.1lims.com/api/materialTypes
2020-05-28 04:09:46.762 | INFO     | api_testing.apis.base_api:info:76 - Status code: 1
2020-05-28 04:09:46.762 | INFO     | api_testing.apis.base_api:info:76 - GET : https://automation.1lims.com/api/testUnits/get/names/1?name=&negelectIsDeleted=0&searchableValue=
2020-05-28 04:09:47.170 | INFO     | api_testing.apis.base_api:info:76 - Status code: 1
2020-05-28 04:09:47.171 | INFO     | api_testing.apis.base_api:info:76 - GET : https://automation.1lims.com/api/testUnits
2020-05-28 04:09:47.682 | INFO     | api_testing.apis.base_api:info:76 - Status code: 1
2020-05-28 04:09:47.682 | INFO     | ui_testing.testcases.base_test:info:129 - Edit order with order no. 20-20-20-20-20-20-20-9075244031 with test_unit b36cd81e2e
2020-05-28 04:09:49.283 | INFO     | ui_testing.pages.base_pages:sleep_small:43 -  Small sleep.
2020-05-28 04:09:54.289 | INFO     | ui_testing.testcases.base_test:info:129 - remove suborder test unit and update it to b36cd81e2e 
2020-05-28 04:09:55.248 | INFO     | ui_testing.pages.base_pages:info:274 -  set test plan : [] for 0 time(s)
2020-05-28 04:09:55.248 | INFO     | ui_testing.pages.base_pages:info:274 -  set test unit : ['b36cd81e2e'] for 1 time(s)
2020-05-28 04:09:55.818 | INFO     | ui_testing.pages.base_pages:sleep_small:43 -  Small sleep.
2020-05-28 04:10:08.430 | INFO     | ui_testing.pages.base_pages:sleep_tiny:39 -  Tiny sleep.
2020-05-28 04:10:10.433 | INFO     | ui_testing.pages.base_pages:info:274 -  save the changes
2020-05-28 04:10:10.478 | INFO     | ui_testing.pages.base_pages:sleep_tiny:39 -  Tiny sleep.
2020-05-28 04:10:12.753 | INFO     | ui_testing.pages.base_pages:sleep_tiny:39 -  Tiny sleep.
2020-05-28 04:10:14.755 | INFO     | ui_testing.testcases.base_test:info:129 - Refresh to make sure that data are saved correctly
2020-05-28 04:10:21.510 | INFO     | ui_testing.pages.base_pages:wait_until_page_is_loaded:531 - wait until page is loaded
2020-05-28 04:10:21.551 | INFO     | ui_testing.testcases.base_test:info:129 - Get suborder data to check it
2020-05-28 04:10:21.669 | INFO     | ui_testing.pages.order_page:get_suborder_data:323 - getting main order data
2020-05-28 04:10:21.839 | INFO     | ui_testing.pages.order_page:get_suborder_data:330 - getting suborders data
2020-05-28 04:10:23.152 | INFO     | ui_testing.testcases.base_test:info:129 - Assert Test units: test units are: b36cd81e2e, and should be: b36cd81e2e
2020-05-28 04:10:23.152 | INFO     | ui_testing.testcases.base_test:info:129 - Getting analysis page to check the data in this child table
2020-05-28 04:10:24.533 | INFO     | ui_testing.pages.base_pages:wait_until_page_is_loaded:531 - wait until page is loaded
2020-05-28 04:10:27.536 | INFO     | ui_testing.pages.base_pages:wait_until_page_is_loaded:531 - wait until page is loaded
2020-05-28 04:10:30.004 | INFO     | ui_testing.pages.base_pages:open_filter_menu:82 -  Open Filter
2020-05-28 04:10:36.445 | INFO     | ui_testing.pages.base_pages:sleep_tiny:39 -  Tiny sleep.
2020-05-28 04:10:43.707 | INFO     | ui_testing.pages.base_pages:sleep_medium:47 -  Medium sleep.
2020-05-28 04:10:55.471 | INFO     | ui_testing.testcases.base_test:info:129 - TearDown.        
ok

----------------------------------------------------------------------
Ran 1 test in 88.607s

OK

```

### CODE REVIEW

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 1]
```diff
@@ -1251,129 +1251,111 @@ def test026_update_suborder_article(self):
...
+                                                 filter_text=suborder_after_refresh['analysis_no'],
+                                                 field_type='text')
+
+        analysis_records = self.analyses_page.get_child_table_data()
+        if len(analysis_records) > 1:
+            test_units = []
+            for a in analysis_records:
+                test_units.append(a['Test Unit'])
+        else:
+            test_units = analysis_records[0]['Test Unit']
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:1352]:
```suggestion
        test_units = []
        for analysis_record in analysis_records:
            test_units.append(analysis_record['Test Unit'])
```

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 2]
```diff
@@ -1251,129 +1251,111 @@ def test026_update_suborder_article(self):
...
+        self.assertIn(suborder_testunits_before_refresh[sub_order_index]['name'], test_units)
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:1353]:
```suggestion
        self.assertIn(suborder_testunits_after_refresh[sub_order_index]['name'], test_units)
```

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 3]
```diff
@@ -1251,129 +1251,111 @@ def test026_update_suborder_article(self):
...
 
-    # def test_test(self):
-    #     import ipdb; ipdb.set_trace()
-    #     suborder_data = self.order_page.get_suborder_data(sub_order_index=3)
-    ### SYNTAX ERROR ###
+        #def test_test(self):
+    #    import ipdb; ipdb.set_trace()
+    #    suborder_data = self.order_page.get_suborder_data(sub_order_index=3)
+    ## SYNTAX ERROR ###
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:1358]: Remove

File: api_testing/apis/orders_api.py [hunk 4]
```diff
@@ -237,6 +239,26 @@ def get_order_with_multiple_sub_orders(self):
             suborder = self.get_suborder_by_order_id(id=order['orderId'])[0]['orders']
             if len(suborder) > 1:
                 return order
+              
+    def get_order_with_feild_name(self, feild):
```
[the remark this answers is not in the document]
Comment (ghada-37) [reply api_testing/apis/orders_api.py:243]: done

File: api_testing/apis/orders_api.py [hunk 5]
```diff
@@ -237,6 +239,26 @@ def get_order_with_multiple_sub_orders(self):
...
+        :param feild: must be in this list ['article', 'materialType','analysis','testPlans','testUnit']
+        :return: order, suborder, suborder_index
+        """
+        orders_data, payload = self.get_all_orders(limit=50)
+        orders = orders_data['orders']
+        for order in orders:
+            suborders_data, a = self.get_suborder_by_order_id(order['id'])
+            suborders = suborders_data['orders']
+            if len(suborders) == 1:
+                if suborders[0][feild]:
```
[the remark this answers is not in the document]
Comment (ghada-37) [reply api_testing/apis/orders_api.py:254]: done

File: api_testing/apis/orders_api.py [hunk 6]
```diff
@@ -237,6 +239,26 @@ def get_order_with_multiple_sub_orders(self):
...
+                    return order, suborders, 0
+            elif len(suborders) == 0:
+                break
```
[the remark this answers is not in the document]
Comment (ghada-37) [reply api_testing/apis/orders_api.py:257]: Right, I updated it

File: api_testing/apis/orders_api.py [hunk 7]
```diff
@@ -237,6 +239,26 @@ def get_order_with_multiple_sub_orders(self):
...
+            else:
+                for i in range(0, len(suborders)-1):
+                    if suborders[i][feild]:
+                        return order, suborders, i
```
[the remark this answers is not in the document]
Comment (ghada-37) [reply api_testing/apis/orders_api.py:261]: Right , I updated it

File: ui_testing/pages/order_page.py [hunk 8]
```diff
@@ -463,8 +471,7 @@ def archive_suborder(self, index, check_pop_up=False):
         archive_element.click()
         self.sleep_tiny()
         if check_pop_up:
-            self.base_selenium.LOGGER.info('confirm archiving')
-            self.base_selenium.click(element='articles:confirm_archive')
+            c
```
[the remark this answers is not in the document]
Comment (ghada-37) [reply ui_testing/pages/order_page.py:474]: sorry for that

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 9]
```diff
@@ -1251,129 +1251,111 @@ def test026_update_suborder_article(self):
...
+        # order's analysis instead of creating new analysis
+        self.order_page.save(save_btn='order:save_btn')
+        self.info('Refresh to make sure that data are saved correctly and analysis no appeared')
+        self.base_selenium.refresh()
+        self.order_page.wait_until_page_is_loaded()
+        self.info('Get suborder data to check it updated correctly')
+        suborder_after_refresh = self.order_page.get_suborder_data()['suborders'][sub_order_index]
+        suborder_testplans_after_refresh = suborder_after_refresh['testplans']
+        self.info('Assert Test plan is: {}, and should be: {}'.format(suborder_testplans_after_refresh, [test_plan]))
+        self.assertEqual(suborder_testplans_after_refresh, [test_plan])
```
[the remark this answers is not in the document]
Comment (ghada-37) [reply ui_testing/testcases/basic_tests/test06_orders.py:1300]: done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 3 above)
Comment (ghada-37) [reply ui_testing/testcases/basic_tests/test06_orders.py:1358]: ok
 I added them as I these lines was on master  master

File: api_testing/apis/test_plan_api.py [hunk 10]
```diff
@@ -209,6 +209,14 @@ def delete_active_testplan(self, id=1):
...
         else:
             return False
+          
+    def get_completed_testplans_with_article_no_and_same_material(self, material_type='Raw Material', articleNo="all"):
+        completed_test_plans = self.get_completed_testplans()
+        test_plans = []
+        for test_plan in completed_test_plans:
+            if test_plan['articleNo'][0] in [articleNo, 'all'] and test_plan['materialType'] == material_type:
+                test_plans.append(test_plan['testPlanName'])
+        return test_plans
```
[the remark this answers is not in the document]
Comment (ghada-37) [reply api_testing/apis/test_plan_api.py:219]: yes, I created this function as other function wasn't merged into master while I was writing code

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 1 above)
Comment (ghada-37) [reply ui_testing/testcases/basic_tests/test06_orders.py]: done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 2 above)
Comment (ghada-37) [reply ui_testing/testcases/basic_tests/test06_orders.py]: updated

**Comment by ghada-37** [comment]
There were bug that when I edit suborder test plan or article by remove old one and add new one, after save both items old and new found 
This bug was solved, but I found sometimes old test plans not removed.

**Comment by ghada-37** [comment]
https://modeso.atlassian.net/browse/LIMS-7913

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 11]
```diff
@@ -1158,129 +1157,76 @@ def test026_update_suborder_article(self):
...
+
+        if not testunits_with_values:
+            self.info(" No test unit with req. material type, so create one")
+            api, testunit_payload = TestUnitAPI().create_quantitative_testunit()
+            self.assertEqual(api['status'], 1)
+            testunit_name = testunit_payload['name']
+            self.info("created test unit name is {}".format(testunit_name))
+        else:
+            testunit_name = random.choice(testunits_with_values)['name']
+            self.info("selected test unit name is {}".format(testunit_name))
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:1191]: abstract it

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 12]
```diff
@@ -1158,129 +1157,76 @@ def test026_update_suborder_article(self):
...
+        self.info("open order edit page")
+        self.orders_page.get_order_edit_page_by_id(order['orderId'])
+        self.order_page.update_suborder(sub_order_index=int(len(sub_order)-1-sub_order_index),
+                                        test_units=[testunit_name], remove_old=True)
+        # checking that when adding new test unit, the newly added test unit is added to the
+        # order's analysis instead of creating new analysis
+        self.order_page.save(save_btn='order:save_btn')
+        self.info('Refresh to make sure that data are saved correctly')
+        self.base_selenium.refresh()
+        self.order_page.wait_until_page_is_loaded()
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:1204]: abstract it as `save_and_wait` method

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 13]
```diff
@@ -1158,129 +1157,76 @@ def test026_update_suborder_article(self):
...
+        self.info('Get suborder data to check it')
+        suborder_after_refresh = \
+            self.order_page.get_suborder_data()['suborders'][int(len(sub_order)-1-sub_order_index)]
+        suborder_testunits_after_refresh = suborder_after_refresh['testunits'][0]['name']
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:1208]: Replace with API call

File: ui_testing/testcases/basic_tests/test06_orders.py [hunk 14]
```diff
@@ -1158,129 +1157,76 @@ def test026_update_suborder_article(self):
...
+        self.analyses_page.apply_filter_scenario(filter_element='analysis_page:analysis_no_filter',
+                                                 filter_text=suborder_after_refresh['analysis_no'],
+                                                 field_type='text')
+        analysis_records = self.analyses_page.get_child_table_data()
+        test_units = []
+        if len(analysis_records) > 1:
+            for analysis_record in analysis_records:
+                test_units.append(analysis_record['Test Unit'])
+        else:
+            test_units = analysis_records[0]['Test Unit']
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_orders.py:1228]:
```suggestion
 test_units = [ analysis_record['Test Unit'] for analysis_record in analysis_records]
```

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 11 above)
Comment (ghada-37) [reply ui_testing/testcases/basic_tests/test06_orders.py:1191]: done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 12 above)
Comment (ghada-37) [reply ui_testing/testcases/basic_tests/test06_orders.py:1204]: done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 14 above)
Comment (ghada-37) [reply ui_testing/testcases/basic_tests/test06_orders.py:1228]: done

File: ui_testing/testcases/basic_tests/test06_orders.py (hunk 13 above)
Comment (ghada-37) [reply ui_testing/testcases/basic_tests/test06_orders.py:1208]: done

**Pull request merged**